libdrgn/program.c

```diff
@@ -655,9 +672,49 @@ static struct drgn_error *get_prstatus_pid(struct drgn_program *prog, const char
 	return NULL;
 }
 
-struct drgn_error *drgn_program_cache_prstatus_entry(struct drgn_program *prog,
-						     const char *data,
-						     size_t size)
+struct drgn_error *drgn_thread_dup_internal(const struct drgn_thread *thread,
+					    struct drgn_thread *ret)
+{
+	struct drgn_error *err = NULL;
+	ret->tid = thread->tid;
+	ret->prog = thread->prog;
+	if (thread->prog->flags &
+	    (DRGN_PROGRAM_IS_LINUX_KERNEL | DRGN_PROGRAM_IS_LIVE)) {
+		drgn_object_init(&ret->thread, thread->prog);
+		err = drgn_object_copy(&ret->thread, &thread->thread);
+		if (err)
+			drgn_object_deinit(&ret->thread);
+	}
+	else {
+		// We don't need a deep copy here since the PRSTATUS notes are cached
+		ret->prstatus = thread->prstatus;
+	}
+	return err;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_thread_dup(const struct drgn_thread *thread, struct drgn_thread **ret)
+{
```

Comment:
Right now for userspace core dumps `struct drgn_thread` doesn't have any mutable state, so this should work just fine and save an allocation, yes. If in the future it _does_ gain any mutable state (which it likely won't since coredumps are inherently static), we'd need to revisit this.

Reply:
Right. I think we do want to avoid the allocation, mainly because then `drgn_thread_destroy` won't need to distinguish between an original `drgn_thread` that was cached in the thread set then returned directly by `drgn_program_find_thread` and a duplicated `drgn_thread` that needs to be freed.

Reply:
That makes sense, but we actually still have some form of the predicament you mentioned. Because we _do_ want to allocate in some cases (for example when debugging the live kernel), `drgn_thread_destroy` _does_ need a way to determine whether or not to `free` to `thread` passed in. This isn't something I had run into issues with because `drgn_thread_dup` isn't presently used anywhere (just `drgn_thread_dup_internal`), but it does need to be addressed. I could just add a `bool` to `struct drgn_thread`, but that feels pretty brittle. Other thoughts on how to approach this?

Reply:
I think what you have now already works: for the kernel, we always allocate, so `drgn_thread_destroy` needs to free. For a userspace core dump, we never allocate, so `drgn_thread_destroy` should be a no-op. And the lifetime rules we've arrived at seem to be:

* `drgn_program_find_thread()` gives you a new `drgn_thread` that should always be passed to `drgn_thread_destroy()` (it is an implementation detail that for userspace core dumps, this is actually a static cached thread and `drgn_thread_destroy` is a no-op, but from the user's point of view, the user owns the `drgn_thread` and is responsible for freeing it).
* `drgn_program_crashed_thread()` lets you borrow a `drgn_thread` that is stored internally. It should not be passed to `drgn_thread_destroy()`.
* `drgn_thread_iterator_next()` similarly lets you borrow a `drgn_thread` that shouldn't be destroyed.
* `drgn_thread_dup()` creates a new thread that should always be passed to `drgn_thread_destroy()` (again, it's an implementation detail that these are no-ops for userspace core dumps).

Reply:
(Side note, I think `drgn_program_deinit` needs to `drgn_thread_destroy(prog->crashed_thread)`, since that would be an allocated thread in the kernel coredump case.)

Reply:
We _don't_ always allocate in the kernel, the existence of `drgn_thread_dup_internal` invalidates this. Consider `Thread_dealloc`. When it is invoked, it should call `drgn_thread_destroy` to destroy its internal `struct drgn_thread`, but obviously `drgn_thread_destroy` should not `free` its argument in this case, since its part of the corresponding `Thread`. Yet in other cases (when `drgn_thread_dup` is invoked when `drgn` is being used as a library), `drgn_thread_destroy` _should_ `free` its argument.

Reply:
Oh, so we actually need `drgn_thread_deinit` that frees the contents of the `drgn_thread` without freeing the `drgn_thread` itself, and `drgn_thread_destroy` which calls that plus frees the `drgn_thread` itself when needed. The `deinit` + `destroy` pattern is used in a few other places where libdrgn users get an opaque pointer but the Python bindings embed it directly: see `drgn_program_deinit` and `drgn_program_destroy`.